### PR TITLE
Move to metadata to `pyproject.toml`

### LINKED ISSUE #56: Switch to setup.cfg for metadata?
Looking at the setuptools documentation, it seems that it's now possible to put  project's metadata in `setup.cfg` (see http://setuptools.readthedocs.io/en/latest/setuptools.html#configuring-setup-using-setup-cfg-files). I don't know when this was introduced, but would it make sense to switch to this approach as the recommended method?

@jaraco Any comments? I don't know when this ability was introduced to setuptools, but it looks really easy to use, and (as far as I'm aware) didn't get as much publicity as maybe it should have...?

### PR DESCRIPTION
**Pull request by di**
Now that PEP 621 support has been merged in https://github.com/pypa/setuptools/pull/2970, we can move project metadata into `pyproject.toml` and get rid of `setup.py` and `setup.cfg`.

The one 'regression' is that external data files are deprecated (https://setuptools.pypa.io/en/latest/userguide/datafiles.html) so support has been removed here.

This should remain a draft PR until https://github.com/pypa/setuptools/issues/2816 is resolved and editable installs work.

Closes #56, closes #57, closes #96.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,5 +1,157 @@
...
+# what's used to render the link text on PyPI.
+[project.urls]  # Optional
+"Homepage" = "https://github.com/pypa/sampleproject"
+"Bug Reports" = "https://github.com/pypa/sampleproject/issues"
+"Funding" = "https://donate.pypi.org"
+"Say Thanks!" = "http://saythanks.io/to/example"
+"Source" = "https://github.com/pypa/sampleproject/"
+
+# To provide executable scripts, use entry points in preference to the
+# "scripts" keyword. Entry points provide cross-platform support and allow
```
Comment (abravalheri) [inline pyproject.toml:139]: To avoid confusion with `project.scripts`, setuptools uses `tool.setuptools.script-files` for the legacy kind of script (which is [explicitly deprecated in the docs now](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setuptools-specific-configuration)).

Maybe we can just skip this advice for the sake of avoiding confusion?
PEP 621 splits general entry-points, `console_scripts` and `gui_scripts` in 3 separated fields. There is a chance the wording *entry point* might be confusing when referring to `scripts` for new pacakgers.

File: pyproject.toml [hunk 2]
```diff
@@ -1,5 +1,157 @@
...
+sample = "sample:main"
+
+[tool.setuptools]
+# If there are data files included in your packages that need to be
+# installed, specify them here.
+package-data = {"sample" = ["*.dat"]}
+
 [build-system]
-# These are the assumed default build requirements from pip:
+# These are the requirements required to build this project
```
Comment (abravalheri) [inline pyproject.toml:154]: How about "needed" to avoid `requirements required`?

```suggestion
# These are the requirements needed to build this project
```

or even

```suggestion
# These are requirements for building this project
```

File: pyproject.toml [hunk 3]
```diff
@@ -1,5 +1,157 @@
+[project]
```
Comment (abravalheri) [inline pyproject.toml:1]: I don't know if it is worthy the trouble, or if it fits the example, but PyPA also maintains a live version of the spec now:

```suggestion
[project]
# A general reference for all the fields available in `[project]` can be found at:
# https://packaging.python.org/en/latest/specifications/declaring-project-metadata/

```

**Review by abravalheri — COMMENTED** [review]
Thank you very much for working on this @di.

From my perspective it looks very good (I just added very minor comments).
I also validated the file with `validate-pyproject` and everything is in shape!

It would be nice though, if possible, to wait at least until the feature stabilises a bit. I think I manage to cover most of unexpected issues in 61.1.1 (there are some very edge-cases left) but we never know... Maybe some users will find other bugs in the following weeks.

**Comment by abravalheri** [comment]
Regarding `data files`, currently there is a proposal https://github.com/pypa/setuptools/issues/3191 for a less problematic and more restricted way of presenting this feature to the users. If you have any thoughts, please feel free to provide some feedback on the discussion, it would be very nice to know what are your impressions.

**Comment by di** [comment]
> Regarding `data files`, currently there is a proposal pypa/setuptools#3191 for a less problematic and more restricted way of presenting this feature to the users. If you have any thoughts, please feel free to provide some feedback on the discussion, it would be very nice to know what are your impressions.

I think I'm OK with removing it entirely from this sample regardless. I feel like overall this is an anti-pattern that we probably don't want to be encouraging for the average use case.

File: pyproject.toml [hunk 4]
```diff
@@ -1,5 +1,160 @@
...
+# Specify which Python versions you support. In contrast to the
+# 'Programming Language' classifiers above, 'pip install' will check this
+# and refuse to install the project if the version does not match. See
+# https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires
+requires-python = ">=3.7"
+
+# This is either text indicating the license for the distribution, or a file
+# that contains the license
+# https://packaging.python.org/en/latest/specifications/core-metadata/#license
+license = {file = "LICENSE.txt"}
```
Comment (xsc27) [inline pyproject.toml:49]: looks like `text is a required field. Error when running `python3 -m build`:
```
configuration error: `project.license` must be valid exactly by one definition (2 matches found):

    - keys:
        'file': {type: string}
      required: ['file']
    - keys:
        'text': {type: string}
      required: ['text']
```suggestion
license = {text = "BSD 3-Clause License", file = "LICENSE.txt"}
```

**Comment by di** [comment]
https://github.com/pypa/setuptools/issues/2816 is now resolved, so moving this out of draft status.

**Comment by mariocj89** [comment]
@di is there anything else blocking this PR? Will this only be landed after setuptools [removes the "beta" banner](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setuptools-specific-configuration) or is it good to land already?

**Comment by di** [comment]
I think I was waiting to not have to require the `tox` prerelease here, but not sure when `tox` is planning to make that release.

Some reviews/approvals on this would be good as well.

**Comment by mariocj89** [comment]
@gaborbernat  do you have an ETA for having a final release of tox 4?

**Comment by gaborbernat** [comment]
No promises. I'm currently blocked by health issues, so probably not this year (assuming no one else picks up development of tox 4).

**Comment by layday** [comment]
Could you restore the `build-system` table here? setuptools falls back on the legacy backend without it.

File: pyproject.toml [hunk 5]
```diff
@@ -1,5 +1,150 @@
...
+  "Topic :: Software Development :: Build Tools",
+
+  # Pick your license as you wish
+  "License :: OSI Approved :: MIT License",
+
+  # Specify the Python versions you support here. In particular, ensure
+  # that you indicate you support Python 3. These classifiers are *not*
+  # checked by "pip install". See instead "python_requires" below.
+  "Programming Language :: Python :: 3",
+  "Programming Language :: Python :: 3.6",
```
Comment (layday) [inline pyproject.toml:91]: rm, we have `requires-python = ">=3.7"`

File: pyproject.toml [hunk 6]
```diff
@@ -1,5 +1,150 @@
...
+  "Programming Language :: Python :: 3.7",
+  "Programming Language :: Python :: 3.8",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
```
Comment (layday) [inline pyproject.toml:95]: Add 3.11?

File: pyproject.toml [hunk 7]
```diff
@@ -1,5 +1,150 @@
...
+  "Programming Language :: Python :: 3 :: Only",
+]
+
+# This field lists other packages that your project depends on to run.
+# Any package you put here will be installed by pip when your project is
+# installed, so they must be valid existing projects.
+#
+# For an analysis of this field vs pip's requirements files see:
+# https://packaging.python.org/discussions/install-requires-vs-requirements/
+dependencies=[
```
Comment (layday) [inline pyproject.toml:105]:
```suggestion
dependencies = [
```

File: pyproject.toml [hunk 8]
```diff
@@ -1,5 +1,150 @@
...
+  "peppercorn"  # Optional
```
Comment (layday) [inline pyproject.toml:106]: Not optional.

File: pyproject.toml [hunk 9]
```diff
@@ -1,5 +1,157 @@
...
+]
+
+# List additional groups of dependencies here (e.g. development
+# dependencies). Users will be able to install these using the "extras"
+# syntax, for example:
+#
+#   $ pip install sampleproject[dev]
+#
+# Similar to `install_requires` above, these must be valid existing
```
[the remark this answers is not in the document]
Comment (layday) [reply pyproject.toml:115]: This is marked as resolved but the change hasn't been applied.

File: pyproject.toml [hunk 10]
```diff
@@ -1,5 +1,150 @@
...
+# https://packaging.python.org/en/latest/specifications/core-metadata/#license
+license = {file = "LICENSE.txt"}
+
+# This field adds keywords for your project which will appear on the
+# project page. What does your project relate to?
+#
+# Note that this is a list of additional keywords, separated
+# by commas, to be used to assist searching for the distribution in a
+# larger catalog.
+keywords= ["sample", "setuptools", "development"]  # Optional
```
Comment (layday) [inline pyproject.toml:54]:
```suggestion
keywords = ["sample", "setuptools", "development"]  # Optional
```

File: pyproject.toml [hunk 11]
```diff
@@ -1,5 +1,150 @@
...
+# List additional groups of dependencies here (e.g. development
+# dependencies). Users will be able to install these using the "extras"
+# syntax, for example:
+#
+#   $ pip install sampleproject[dev]
+#
+# Similar to `install_requires` above, these must be valid existing
+# projects.
+[project.optional-dependencies]
+test = ["coverage"] # Optional
```
Comment (layday) [inline pyproject.toml:118]:
```suggestion
[project.optional-dependencies] # Optional
test = ["coverage"]
```

**Comment by layday** [comment]
I’m not sure I understand why you want to omit the build system table and therefore block yourself on tox.

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Comment by pfmoore** [comment]
Given that we *currently* have a `build-system` section, removing it and then claiming that we need to block this until tox can handle the removal, seems a little unreasonable. Why not move this forward with `build-system`, and then follow up with a PR removing that section, if you really care that strongly?

I should also note that PEP 517 notes:

> If the `pyproject.toml` file is absent, or the build-backend key is missing, the source tree is not using this specification, and tools should revert to the legacy behaviour of running setup.py (either directly, or by implicitly invoking the setuptools.build_meta:__legacy__ backend).

In particular, this means that the *legacy* setuptools backend will be used in the absence of `build-system`. And I definitely don't think that's something we want to encourage people to do.

**Comment by di** [comment]
Fair points @pfmoore. I think all outstanding comments/suggestions have been resolved.

**Comment by layday** [comment]
Well, I'll also note that the provision in PEP 518 as I understand it was added to ease the transition, not to elevate setuptools above other backends; the recommendation for new projects has been to include it.  In addition, omitting the build system table subtly changes the behaviour of setuptools, placing the project root on the Python path.

**Comment by layday** [comment]
Who can approve this PR?

File: tox.ini [hunk 12]
```diff
@@ -3,8 +3,6 @@
 #
 #  It's meant to show the use of:
 #
-#  - check-manifest
-#     confirm items checked into vcs are in your sdist
```
Comment (pfmoore) [inline tox.ini:7]: Removing `check-manifest` seems like it's an independent change here.

I have no great love for including it (my original intention for this project was that it be *minimal*) but I think it should be removed in a separate PR, keeping this PR for the move to `pyproject.toml`.

File: setup.py [hunk 13]
```diff
@@ -1,181 +0,0 @@
...
-    # installed, specify them here.
-    package_data={  # Optional
-        "sample": ["package_data.dat"],
-    },
-    # Although 'package_data' is the preferred approach, in some case you may
-    # need to place data files outside of your packages. See:
-    # http://docs.python.org/distutils/setupscript.html#installing-additional-files
-    #
-    # In this case, 'data_file' will be installed into '<sys.prefix>/my_data'
-    data_files=[("my_data", ["data/data_file"])],  # Optional
```
Comment (pfmoore) [inline setup.py:153]: Removing the inclusion of `data/data_file` (sorry, I appear not to be able to comment on the removed file itself) seems independent of switching to `pyproject.toml`. I'm fine with *not* showing an example of non-package data, as that's not recommended practice these days, but it should be in a separate PR.

File: tox.ini [hunk 14]
```diff
@@ -28,17 +26,15 @@ isolated_build = true
...
+    build
+    twine
 commands =
-    check-manifest --ignore 'tox.ini,tests/**'
-    # This repository uses a Markdown long_description, so the -r flag to
-    # `setup.py check` is not needed. If your project contains a README.rst,
-    # use `python setup.py check -m -r -s` instead.
-    python setup.py check -m -s
+    python -m build
+    python -m twine check dist/*
```
Comment (pfmoore) [inline tox.ini:37]: This whole section of the PR is unrelated to moving to `pyproject.toml` (see previous comments to the same effect).

**Review by pfmoore — CHANGES_REQUESTED** [review]
A few comments related to making this PR be *solely* about moving metadata into `pyproject.toml`. Can the unrelated changes be moved to their own PR(s), please?

File: tox.ini (hunk 14 above)
Comment (layday) [reply tox.ini:37]: Well, there's no `setup.py` to invoke. What would you like to see here?

Comment (pfmoore) [reply tox.ini:37]: Removing `setup.py` isn't necessary - moving the metadata out is the point of this PR, not removing `setup.py` (or moving off setuptools in general).

Comment (layday) [reply tox.ini:37]: I'm not sure that I agree with splitting the concomitant configuration changes into separate PRs, but IIUC what you're suggesting is to have a `setup.py` with just:

```py
from setuptools import setup

setup()
```

Is that right?

File: setup.py (hunk 13 above)
Comment (layday) [reply setup.py:153]: @abravalheri How should this be represented in `pyproject.toml` - does it need to be a list of lists of lists?

```toml
[tool.setuptools]
data-files = [
  ["my_data", ["data/data_file"]],
]
```

File: tox.ini (hunk 12 above)
Comment (di) [reply tox.ini]: I can't recall why I removed this, it must have not been working with `pyproject.toml` at the time (note that the CI run prior to the commit that removed this failed, but the logs are no longer available), but this appears to work with the latest version of `check-mainifest`, so I reverted it in 8945e5e1426c97c64ce1748d9bd69b1e038c7e30.

File: tox.ini (hunk 14 above)
Comment (di) [reply tox.ini:41]: I didn't see any reason to keep an empty `setup.py` around. We're still using `setuptools` as the build backend, so it shouldn't make a difference, right?

File: setup.py (hunk 13 above)
Comment (di) [reply setup.py:153]: There's some previous discussion about this [here](https://github.com/pypa/sampleproject/pull/166#issuecomment-1079727615) and [here](https://github.com/pypa/sampleproject/pull/166#issuecomment-1080895022). IIRC this needs https://github.com/pypa/setuptools/issues/3191 in order to work?

File: tox.ini (hunk 14 above)
Comment (pfmoore) [reply tox.ini:41]: Agreed, if having `setup.py` isn't needed to keep the functionality the same, then drop it.

Also, if we don't want to include a `data-files` example, and/or an example of using `check-manifest`, by all means drop them. But in separate PRs. If thise PRs get merged before this one, we don't even need to think about how to represent the functionality in `pyproject.toml`.

All I'm asking for here is for each PR to do one well-defined thing, I'm not arguing that we need to keep all the examples we currently have. (TBH, this project already has *way* more in it than I'd prefer - if it were down to me, I'd remove the whole tox configuration, for example).

File: README.md [hunk 15]
```diff
@@ -12,7 +12,7 @@ recommendations for version control, documentation, or testing.
 
 [The source for this project is available here][src].
 
-Most of the configuration for a Python project is done in the `setup.py` file,
+Most of the configuration for a Python project is done in the `pyproject.toml` file,
```
Comment (pfmoore) [inline README.md:15]:
```suggestion
The metadata for a Python project is defined in the `pyproject.toml` file,
```

Comment (pfmoore) [reply README.md:15]: Wording it as suggested avoids the question of whether build backend configuration should/must be done in `pyproject.toml`.

File: pyproject.toml (hunk 1 above)
Comment (pfmoore) [reply pyproject.toml:139]: Yes, this whole comment needs either re-wording or removing.

We shouldn't be mentioning the setuptools-specific "scripts" keyword at all here.

File: setup.py (hunk 13 above)
Comment (pfmoore) [reply setup.py:153]: Frankly if setuptools maintainers have reservations about allowing this functionality (as https://github.com/pypa/setuptools/issues/3191 suggests) we'd probably be better just removing it as an example altogether.

**Comment by pfmoore** [comment]
Does https://github.com/pypa/sampleproject/pull/181 help? It gets rid of the various deprecated features that are causing a lot of the debate here.

**Comment by pfmoore** [comment]
I've fixed up this PR after merging #181 into main. Hope I did it right!

**Review by pfmoore — APPROVED** [review]

**Comment by pfmoore** [comment]
It feels a bit off to merge based on my own approval, so @di I'll wait for your OK (or you can merge if you're happy).

**Comment by pfmoore** [comment]
Hmm, we may have conflicted here. @di I think your force-push blew away some changes I made :-(

**Comment by di** [comment]
Oops, sorry @pfmoore, I didn't see your changes here before I rebased on main!

**Comment by di** [comment]
@pfmoore I think I restored all your changes here, if it looks good to you, please merge!

**Comment by pfmoore** [comment]
No worries, I normally do PRs from my own fork, messing about with branches in the main repo is new to me so I didn't know the protocols :-)

**Comment by pfmoore** [comment]
Looks good to me. I'd dropped the `tox.ini` changes because they are unrelated to the move of the metadata, but I'm not going to make a big deal about it either way.

I'll merge this once checks are complete.

**Comment by pfmoore** [comment]
lol looks like they completed while I was typing that!

File: setup.cfg [hunk 16]
```diff
@@ -1,4 +0,0 @@
-[metadata]
```
Comment (webknjaz) [inline setup.cfg:1]: FTR this broke the linkcheck in PyPUG
